File: Week3/ArrayProblem/SortArrayWithSquares.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class SortArrayWithSquares{
/*
 * Given a sorted array A containing N integers both positive and negative.
 * You need to create another array containing the squares of all the elements in A and return it in non-decreasing order.
 * https://www.interviewbit.com/problems/sort-array-with-squares/
 * T(c) -> O(nlogn), S(c) -> O(n)
 */
    static ArrayList<Integer> sortWithSquares(ArrayList<Integer> A) {
        ArrayList<Integer> result = new ArrayList<>();
        
        for(int i=0;i<A.size();i++)
        {
            int num = A.get(i);
            int square = num * num;
            result.add(i,square);
        }
        Collections.sort(result);
        return result;
    }

    /*
     * check from both end of array and insert in result accordingly
     * T(c) -> O(n), S(c) -> O(n)
     */
    static ArrayList<Integer> sortWithSquaresEfficient(ArrayList<Integer> A) {
        int n = A.size();
        ArrayList<Integer> result = new ArrayList<>(Collections.nCopies(n,0));
        
        int left = 0;
        int right = n-1;
        int i = n-1;
        
        while(left <= right)
        {
            int leftSquare = A.get(left) * A.get(left);
            int rightSqaure = A.get(right) * A.get(right);
            if(rightSqaure > leftSquare)
            {
                result.set(i--,rightSqaure);
                right--;
            }
            else{
                result.set(i--,leftSquare);
                left++;
            }
        }

        return result;
    }

    public static void main(String[] args) {
        ArrayList<Integer> arr = new ArrayList<>(List.of(-6, -3, -1, 2, 4, 5));
        System.out.println(sortWithSquares(arr));
        System.out.println(sortWithSquaresEfficient(arr));
    }
}

File: Week3/Python/ArrayProblems/SortArrayWithSquares.py
"""
    Given a sorted array A containing N integers both positive and negative.
    You need to create another array containing the squares of all the elements in A and return it in non-decreasing order.
    Try to this in O(N) time.

    https://www.interviewbit.com/problems/sort-array-with-squares/
"""

def solve(A):
    """ T(c) -> O(n), S(c) -> O(1) """

    # result array
    result = [0] * len(A)

    left = 0
    right = len(A) - 1
    k = len(result) - 1

    # using 2 pointer approach
    while left <= right:
        # place elements from greater to smaller element
        if((A[left] ** 2) > (A[right] ** 2)):
            result[k] = A[left] ** 2
            left = left+1
        else:
            result[k] = A[right] ** 2
            right = right-1
        k = k-1
    
    return result


arr = [-6, -3, -1, 2, 4, 5]
print(arr)
result = solve(arr)
print(result)
